File: ios/Classes/FlutterInhousePlugin.swift
import Flutter
import UIKit
import InhouseTrackingSDK

public class FlutterInhousePlugin: NSObject, FlutterPlugin, FlutterStreamHandler {
    private var eventSink: FlutterEventSink?
    private var methodChannel: FlutterMethodChannel?
    private var eventChannel: FlutterEventChannel?
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = FlutterInhousePlugin()
        
        let methodChannel = FlutterMethodChannel(
            name: "flutter_inhouse",
            binaryMessenger: registrar.messenger()
        )
        instance.methodChannel = methodChannel
        registrar.addMethodCallDelegate(instance, channel: methodChannel)
        
        let eventChannel = FlutterEventChannel(
            name: "flutter_inhouse/callbacks",
            binaryMessenger: registrar.messenger()
        )
        instance.eventChannel = eventChannel
        eventChannel.setStreamHandler(instance)
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "initialize":
            handleInitialize(call, result: result)
        case "onAppResume":
            handleOnAppResume(call, result: result)
        case "onNewURL":
            handleOnNewURL(call, result: result)
        case "trackAppOpen":
            handleTrackAppOpen(call, result: result)
        case "trackSessionStart":
            handleTrackSessionStart(call, result: result)
        case "trackShortLinkClick":
            handleTrackShortLinkClick(call, result: result)
        case "getInstallReferrer":
            handleGetInstallReferrer(call, result: result)
        case "fetchInstallReferrer":
            handleFetchInstallReferrer(call, result: result)
        case "resetFirstInstall":
            handleResetFirstInstall(call, result: result)
        case "getFingerprint":
            // iOS doesn't support fingerprinting, return empty string
            result("")
        case "getFingerprintId":
            // iOS doesn't support fingerprinting, return empty string
            result("")
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    // MARK: - FlutterStreamHandler
    
    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        return nil
    }
    
    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.eventSink = nil
        return nil
    }
    
    // MARK: - Method Handlers
    
    private func handleInitialize(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let projectToken = args["projectToken"] as? String,
              let tokenId = args["tokenId"] as? String,
              let shortLinkDomain = args["shortLinkDomain"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENTS", message: "Missing required arguments", details: nil))
            return
        }
        
        let serverUrl = args["serverUrl"] as? String
        let enableDebugLogging = args["enableDebugLogging"] as? Bool ?? false
        
        DispatchQueue.main.async {
            InhouseTrackingSDK.shared.initialize(
                projectId: tokenId,
                projectToken: projectToken,
                shortLinkDomain: shortLinkDomain,
                serverUrl: serverUrl,
                enableDebugLogging: enableDebugLogging
            ) { [weak self] callbackType, data in
                self?.sendEvent(callbackType: callbackType, data: data)
            }
            result("SDK initialized successfully")
        }
    }
    
    private func handleOnAppResume(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        DispatchQueue.main.async {
            InhouseTrackingSDK.shared.onAppResume()
            result(nil)
        }
    }
    
    private func handleOnNewURL(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let urlString = args["url"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENTS", message: "Missing URL argument", details: nil))
            return
        }
        
        DispatchQueue.main.async {
            let url = URL(string: urlString)
            InhouseTrackingSDK.shared.onNewURL(url)
            result(nil)
        }
    }
    
    private func handleTrackAppOpen(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as? [String: Any]
        let shortLink = args?["shortLink"] as? String
        
        DispatchQueue.main.async {
            InhouseTrackingSDK.shared.trackAppOpen(shortLink: shortLink) { responseJson in
                result(responseJson)
            }
        }
    }
    
    private func handleTrackSessionStart(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as? [String: Any]
        let shortLink = args?["shortLink"] as? String
        
        DispatchQueue.main.async {
            InhouseTrackingSDK.shared.trackSessionStart(shortLink: shortLink) { responseJson in
                result(responseJson)
            }
        }
    }
    
    private func handleTrackShortLinkClick(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let shortLink = args["shortLink"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENTS", message: "Missing shortLink argument", details: nil))
            return
        }
        
        let deepLink = args["deepLink"] as? String
        
        DispatchQueue.main.async {
            InhouseTrackingSDK.shared.trackShortLinkClick(shortLink: shortLink, deepLink: deepLink) { responseJson in
                result(responseJson)
            }
        }
    }
    
    private func handleGetInstallReferrer(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        DispatchQueue.main.async {
            let installReferrer = InhouseTrackingSDK.shared.getInstallReferrer()
            result(installReferrer ?? "")
        }
    }
    
    private func handleFetchInstallReferrer(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        DispatchQueue.main.async {
            InhouseTrackingSDK.shared.fetchInstallReferrer { referrer in
                result(referrer ?? "")
            }
        }
    }
    
    private func handleResetFirstInstall(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        DispatchQueue.main.async {
            InhouseTrackingSDK.shared.resetFirstInstall()
            result(nil)
        }
    }
    
    // MARK: - Helper Methods
    
    private func sendEvent(callbackType: String, data: String) {
        guard let eventSink = self.eventSink else { return }
        
        let eventData: [String: Any] = [
            "callbackType": callbackType,
            "data": data
        ]
        
        DispatchQueue.main.async {
            eventSink(eventData)
        }
    }
} 

File: android/src/main/kotlin/co/tryinhouse/flutter/FlutterInhousePlugin.kt
package co.tryinhouse.flutter

import android.content.Context
import android.os.Handler
import android.os.Looper
import android.util.Log
import androidx.annotation.NonNull
import co.tryinhouse.android.TrackingSDK
import com.thumbmarkjs.thumbmark_android.Thumbmark
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/** FlutterInhousePlugin */
class FlutterInhousePlugin: FlutterPlugin, MethodCallHandler, EventChannel.StreamHandler {
    /// The MethodChannel that will the communication between Flutter and native Android
    ///
    /// This local reference serves to register the plugin with the Flutter Engine and unregister it
    /// when the Flutter Engine is detached from the Activity
    private lateinit var methodChannel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private var context: Context? = null
    private var eventSink: EventChannel.EventSink? = null

    override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        context = flutterPluginBinding.applicationContext
        
        methodChannel = MethodChannel(flutterPluginBinding.binaryMessenger, "flutter_inhouse")
        methodChannel.setMethodCallHandler(this)
        
        eventChannel = EventChannel(flutterPluginBinding.binaryMessenger, "flutter_inhouse/callbacks")
        eventChannel.setStreamHandler(this)
    }

    override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
        when (call.method) {
            "initialize" -> handleInitialize(call, result)
            "onAppResume" -> handleOnAppResume(call, result)
            "onNewURL" -> {
                // Android doesn't need onNewURL, just return success
                result.success(null)
            }
            "trackAppOpen" -> handleTrackAppOpen(call, result)
            "trackSessionStart" -> handleTrackSessionStart(call, result)
            "trackShortLinkClick" -> handleTrackShortLinkClick(call, result)
            "getInstallReferrer" -> handleGetInstallReferrer(call, result)
            "fetchInstallReferrer" -> handleFetchInstallReferrer(call, result)
            "resetFirstInstall" -> handleResetFirstInstall(call, result)
            "getFingerprint" -> handleGetFingerprint(call, result)
            "getFingerprintId" -> handleGetFingerprintId(call, result)
            else -> result.notImplemented()
        }
    }

    override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        context = null
    }

    // MARK: - EventChannel.StreamHandler

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }

    // MARK: - Method Handlers

    private fun handleInitialize(call: MethodCall, result: Result) {
        try {
            val projectToken = call.argument<String>("projectToken")
            val tokenId = call.argument<String>("tokenId")
            val shortLinkDomain = call.argument<String>("shortLinkDomain")
            val serverUrl = call.argument<String>("serverUrl")
            val enableDebugLogging = call.argument<Boolean>("enableDebugLogging") ?: false

            if (projectToken == null || tokenId == null || shortLinkDomain == null) {
                result.error("INVALID_ARGUMENTS", "Missing required arguments", null)
                return
            }

            val context = this.context
            if (context == null) {
                result.error("CONTEXT_NULL", "Android context is null", null)
                return
            }

            Log.d("FlutterInhousePlugin", "initialize called with projectToken=$projectToken, tokenId=$tokenId, shortLinkDomain=$shortLinkDomain, serverUrl=$serverUrl, enableDebugLogging=$enableDebugLogging")

            TrackingSDK.getInstance().initialize(
                context = context,
                projectToken = projectToken,
                tokenId = tokenId,
                shortLinkDomain = shortLinkDomain,
                serverUrl = serverUrl ?: "https://api.tryinhouse.co",
                enableDebugLogging = enableDebugLogging
            ) { callbackType, jsonData ->
                Log.d("FlutterInhousePlugin", "SDK callback: callbackType=$callbackType, data=$jsonData")
                sendEvent(callbackType, jsonData)
            }
            result.success("SDK initialized successfully")
        } catch (e: Exception) {
            Log.e("FlutterInhousePlugin", "Initialization error", e)
            result.error("INITIALIZATION_ERROR", e.message, null)
        }
    }

    private fun handleOnAppResume(call: MethodCall, result: Result) {
        try {
            TrackingSDK.getInstance().onAppResume()
            result.success(null)
        } catch (e: Exception) {
            result.error("APP_RESUME_ERROR", e.message, null)
        }
    }

    private fun handleTrackAppOpen(call: MethodCall, result: Result) {
        try {
            val shortLink = call.argument<String>("shortLink")
            TrackingSDK.getInstance().trackAppOpen(shortLink) { responseJson ->
                result.success(responseJson)
            }
        } catch (e: Exception) {
            result.error("TRACK_APP_OPEN_ERROR", e.message, null)
        }
    }

    private fun handleTrackSessionStart(call: MethodCall, result: Result) {
        try {
            val shortLink = call.argument<String>("shortLink")
            TrackingSDK.getInstance().trackSessionStart(shortLink) { responseJson ->
                result.success(responseJson)
            }
        } catch (e: Exception) {
            result.error("TRACK_SESSION_START_ERROR", e.message, null)
        }
    }

    private fun handleTrackShortLinkClick(call: MethodCall, result: Result) {
        try {
            val shortLink = call.argument<String>("shortLink")
            val deepLink = call.argument<String>("deepLink")
            
            if (shortLink == null) {
                result.error("INVALID_ARGUMENTS", "Missing shortLink argument", null)
                return
            }
            
            TrackingSDK.getInstance().trackShortLinkClick(shortLink, deepLink) { responseJson ->
                result.success(responseJson)
            }
        } catch (e: Exception) {
            result.error("TRACK_SHORT_LINK_CLICK_ERROR", e.message, null)
        }
    }

    private fun handleGetInstallReferrer(call: MethodCall, result: Result) {
        try {
            val installReferrer = TrackingSDK.getInstance().getInstallReferrer()
            result.success(installReferrer ?: "")
        } catch (e: Exception) {
            result.error("GET_INSTALL_REFERRER_ERROR", e.message, null)
        }
    }

    private fun handleFetchInstallReferrer(call: MethodCall, result: Result) {
        try {
            TrackingSDK.getInstance().fetchInstallReferrer { referrer ->
                result.success(referrer ?: "")
            }
        } catch (e: Exception) {
            result.error("FETCH_INSTALL_REFERRER_ERROR", e.message, null)
        }
    }

    private fun handleResetFirstInstall(call: MethodCall, result: Result) {
        try {
            TrackingSDK.getInstance().resetFirstInstall()
            result.success(null)
        } catch (e: Exception) {
            result.error("RESET_FIRST_INSTALL_ERROR", e.message, null)
        }
    }

    private fun handleGetFingerprint(call: MethodCall, result: Result) {
        try {
            val context = this.context
            if (context == null) {
                result.success("")
                return
            }
            
            val fingerprint = Thumbmark.fingerprint(context)
            result.success(fingerprint?.toString() ?: "")
        } catch (e: Exception) {
            Log.e("FlutterInhousePlugin", "Error getting fingerprint", e)
            result.success("")
        }
    }

    private fun handleGetFingerprintId(call: MethodCall, result: Result) {
        try {
            val context = this.context
            if (context == null) {
                result.success("")
                return
            }
            
            val algorithm = call.argument<String>("algorithm")
            val fingerprintId = if (algorithm != null) {
                Thumbmark.id(algorithm, context)
            } else {
                Thumbmark.id(context)
            }
            result.success(fingerprintId ?: "")
        } catch (e: Exception) {
            Log.e("FlutterInhousePlugin", "Error getting fingerprint ID", e)
            result.success("")
        }
    }

    // MARK: - Helper Methods

    private fun sendEvent(callbackType: String, data: String) {
        val eventData = mapOf(
            "callbackType" to callbackType,
            "data" to data
        )
        
        Handler(Looper.getMainLooper()).post {
            eventSink?.success(eventData)
        }
    }
} 